File: 2024-Practica 2-DivideDecrease/Factorial/Java/Factorial.java
public class Factorial {
  //~ Divide and conquer
  public static long factorialDivideConquer(int n) {
    if (n < 0) {
      throw new IllegalArgumentException("El factorial no está definido para números negativos");
    }
    return divideConquer(1, n);
  }

  private static long divideConquer(int ini, int end) {
    if (ini == end) {
      return ini;
    } else {
      int mid = (ini + end) / 2;
      return divideConquer(ini, mid) * divideConquer(mid + 1, end);
    }
  }

  //~ Decrease and conquer
  public static long factorialDecreaseConquer(int n) {
    if (n < 0) {
      throw new IllegalArgumentException("El factorial no está definido para números negativos");
    }
    return decreaseConquer(1, n);
  }

  private static long decreaseConquer(long acc, int n) {
    if (n == 1) {
      return acc;
    } else {
      return decreaseConquer(acc * n, n - 1);
    }
  }

  public static void main(String[] args) {
    System.out.println("Factorial usando Divide and Conquer:");
    System.out.println(Factorial.factorialDivideConquer(5));

    System.out.println("Factorial usando Decrease and Conquer:");
    System.out.println(Factorial.factorialDecreaseConquer(5));
  }
}

File: 2024-Practica 2-DivideDecrease/Factorial/Haskell/Factorial.hs
factorialDivideAndConquer :: Integer -> Integer
factorialDivideAndConquer 0 = 1
factorialDivideAndConquer n = divide 1 n

divide :: Integer -> Integer -> Integer
divide n m
  | n == m     = n
  | otherwise  = let mid = (n + m) `div` 2
                 in divide n mid * divide (mid + 1) m

factorialDecreaseAndConquer :: Integer -> Integer
factorialDecreaseAndConquer 0 = 1
factorialDecreaseAndConquer n = decrease 1 n

decrease :: Integer -> Integer -> Integer
decrease acum 1 = acum
decrease acum n = decrease (acum * n) (n - 1)

main :: IO ()
main = do
    putStrLn "Factorial usando Divide and Conquer:"
    print $ factorialDivideAndConquer 5
    putStrLn "Factorial usando Decrease and Conquer:"
    print $ factorialDecreaseAndConquer 5
